### PR TITLE
[CPDLP-3775] NPQ Post Separation Cleanup - Remove feature flag

### PR DESCRIPTION
**Pull request by mooktakim**
### Context

Ticket: https://dfedigital.atlassian.net/browse/CPDLP-3775

Rebased: https://github.com/DFE-Digital/npq-registration/pull/2034

### Changes proposed in this pull request

* Updated `Application.ecf_id` to auto generated UUID
* Updated `User.ecf_id` to auto generated UUID
* Removed feature flag from specs
* Deleted some ecf endpoint services that no longer needed

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -28,14 +28,8 @@ class User < ApplicationRecord
...
   validates :uid, uniqueness: { allow_blank: true }
-  # TODO: add constraints into the DB after separation
-  validates :ecf_id, presence: true, if: -> { Feature.ecf_api_disabled? }
-  validates :ecf_id, uniqueness: { allow_blank: true }
-
-  # TODO: remove this and add default: "gen_random_uuid()" in the DB after separation
-  before_validation do
-    self.ecf_id ||= SecureRandom.uuid if Feature.ecf_api_disabled? && ecf_id.blank?
-  end
+  validates :uid, inclusion: { in: ->(user) { [user.uid_was] } }, on: :npq_separation, if: -> { uid_was.present? }
```
Comment (cwrw) [inline app/models/user.rb:31]: I think this is a rebase error, can we remove please

File: db/migrate/20241203191058_auto_gen_uuid_ecf_id_for_applications.rb [hunk 2]
```diff
@@ -0,0 +1,6 @@
+class AutoGenUuidEcfIdForApplications < ActiveRecord::Migration[7.1]
+  def change
+    change_column_default :applications, :ecf_id, from: nil, to: "gen_random_uuid()"
```
Comment (cwrw) [inline db/migrate/20241203191058_auto_gen_uuid_ecf_id_for_applications.rb:3]: worth checking on all envs that this value is present incase :) same with users

File: spec/features/journeys/happy_paths/previously_received_targeted_delivery_funding_spec.rb [hunk 3]
```diff
@@ -119,7 +120,7 @@ def run_scenario(js:)
       "cohort_id" => Cohort.current.id,
       "course_id" => Course.find_by(identifier: "npq-senior-leadership").id,
       "schedule_id" => nil,
-      "ecf_id" => nil,
+      "ecf_id" => Application.last.ecf_id,
```
Comment (cwrw) [inline spec/features/journeys/happy_paths/previously_received_targeted_delivery_funding_spec.rb:123]: will this be flakey? can we get a more accurate application here and everywhere else?

File: spec/support/helpers/api_mocking.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+def mock_previous_funding(previously_funded:, previously_received_targeted_funding_support: false)
```
Comment (leandroalemao) [inline spec/support/helpers/api_mocking.rb:1]: This can be removed can't it?

**Review by leandroalemao — CHANGES_REQUESTED** [review]

File: spec/models/user_spec.rb [hunk 5]
```diff
@@ -34,43 +34,11 @@
 
   describe "validations" do
     it { is_expected.to validate_presence_of(:full_name).with_message("Enter a full name") }
-    it { is_expected.to validate_presence_of(:email).with_message("Enter an email address") }
+    it { is_expected.to validate_presence_of(:email).on(:npq_separation).with_message("Enter an email address") }
```
Comment (cwrw) [inline spec/models/user_spec.rb:37]: we removed the on separation logic, worth updating with main

File: spec/models/application_spec.rb [hunk 6]
```diff
@@ -274,7 +241,7 @@
 
   describe "versioning", :versioning do
     context "when changing versioned fields" do
-      let(:application) { create(:application, lead_provider_approval_status: "pending", participant_outcome_state: nil) }
+      let(:application) { create(:application, lead_provider_approval_status: "pending", participant_outcome_state: nil, funded_place: false) }
```
Comment (cwrw) [inline spec/models/application_spec.rb:244]: funded place is `nil` if the application is pending?

File: spec/support/helpers/api_mocking.rb (hunk 4 above)
Comment (cwrw) [inline spec/support/helpers/api_mocking.rb:1]: I think this got deleted? don't think it's used anywhere?

File: spec/services/applications/accept_spec.rb [hunk 7]
```diff
@@ -19,8 +19,8 @@
     let(:course_group) { create(:course_group, name: "leadership") }
     let(:course) { create(:course, :senior_leadership, course_group:) }
     let(:lead_provider) { create(:lead_provider) }
-    let(:cohort) { create(:cohort, :current) }
-    let(:cohort_next) { create(:cohort, :next) }
+    let(:cohort) { create(:cohort, :current, funding_cap: false) }
```
Comment (cwrw) [inline spec/services/applications/accept_spec.rb:22]: should the default be false 🤔 why did this start failing as it was ok before?

File: spec/services/participants/change_schedule_spec.rb [hunk 8]
```diff
@@ -3,7 +3,7 @@
 require "rails_helper"
 
 RSpec.describe Participants::ChangeSchedule, type: :model do
-  let(:cohort) { create(:cohort, :current) }
+  let(:cohort) { create(:cohort, :current, funding_cap: false) }
```
Comment (cwrw) [inline spec/services/participants/change_schedule_spec.rb:6]: same here why did this start failing?

File: spec/serializers/api/applications_csv_serializer_spec.rb [hunk 9]
```diff
@@ -21,7 +22,7 @@
     let(:first_row) { rows.first.to_hash.symbolize_keys }
 
     it { expect(rows.count).to eq(applications.count) }
-    it { expect(first_row.values).to all(be_present) }
+    it { expect(first_row.except(:ineligible_for_funding_reason).values).to all(be_present) }
```
Comment (cwrw) [inline spec/serializers/api/applications_csv_serializer_spec.rb:25]: why did we need to exclude this?

File: spec/lib/services/get_an_identity_service/webhooks/user_updated_processor_spec.rb [hunk 10]
```diff
@@ -38,6 +38,26 @@
     let(:new_trn) { rand(1_000_000..9_999_999).to_s }
     let(:new_trn_status) { "found" }
 
+    it "updates user data and sends it to ECF" do
```
Comment (cwrw) [inline spec/lib/services/get_an_identity_service/webhooks/user_updated_processor_spec.rb:41]: I think this should go as well?

**Review by cwrw — COMMENTED** [review]
thanks Mook, a few comments, also we still have a few places we are calling `ecf_api_disabled` from if you could remove them please

File: spec/features/journeys/happy_paths/previously_received_targeted_delivery_funding_spec.rb (hunk 3 above)
Comment (mooktakim) [reply spec/features/journeys/happy_paths/previously_received_targeted_delivery_funding_spec.rb:123]: Ah yes, will fix those

File: spec/support/helpers/api_mocking.rb (hunk 4 above)
Comment (mooktakim) [reply spec/support/helpers/api_mocking.rb:1]: Ah yes, not used anymore

File: spec/serializers/api/applications_csv_serializer_spec.rb (hunk 9 above)
Comment (mooktakim) [reply spec/serializers/api/applications_csv_serializer_spec.rb:25]: @cwrw when there is no reason this column is nil, so all values are not set. I could set it to something maybe or change the entire expect line

File: db/migrate/20241203191058_auto_gen_uuid_ecf_id_for_applications.rb (hunk 2 above)
Comment (mooktakim) [reply db/migrate/20241203191058_auto_gen_uuid_ecf_id_for_applications.rb:3]: Updated migration to generate ecf_id if nil for User and Application

File: spec/models/user_spec.rb (hunk 5 above)
Comment (mooktakim) [reply spec/models/user_spec.rb:37]: removed

File: spec/models/application_spec.rb (hunk 6 above)
Comment (mooktakim) [reply spec/models/application_spec.rb:244]: removed, this isn't needed

File: spec/services/participants/change_schedule_spec.rb (hunk 8 above)
Comment (mooktakim) [reply spec/services/participants/change_schedule_spec.rb:6]: removed

File: spec/lib/services/get_an_identity_service/webhooks/user_updated_processor_spec.rb (hunk 10 above)
Comment (mooktakim) [reply spec/lib/services/get_an_identity_service/webhooks/user_updated_processor_spec.rb:41]: I think we just change the name, it doesn't send to ECF, only updates

File: spec/serializers/api/applications_csv_serializer_spec.rb (hunk 9 above)
Comment (mooktakim) [reply spec/serializers/api/applications_csv_serializer_spec.rb]: Fixed

File: spec/services/applications/accept_spec.rb (hunk 7 above)
Comment (mooktakim) [reply spec/services/applications/accept_spec.rb]: Reverted this, should pass now

**Review by ethax-ross — APPROVED** [review]
Looks good to me 👏 are we going to rename `ecf_id` to something else now 🤔

**Review by cwrw — APPROVED** [review]
thanks @mooktakim!

**Review by leandroalemao — APPROVED** [review]
All good tks @mooktakim 👍

**Pull request merged**